lib/src/local_working_copy.rs

```diff
@@ -104,19 +107,117 @@ use crate::working_copy::WorkingCopy;
 use crate::working_copy::WorkingCopyFactory;
 use crate::working_copy::WorkingCopyStateError;
 
+/// The executable bit for a filetype, potentially ignored.
+///
+/// On Windows there is no executable bit, so this will always be Ignore. On
+/// Unix it will usually be Exec(true|false), but may be ignored by a config
+/// value or by a check we run when we load the config if not specified.
+#[derive(Debug, Clone, Copy)]
+pub enum ExecFlag {
+    Exec(bool),
+    Ignore,
+}
+// Note: cannot derive Eq or PartialEq since a == b == c does not imply a == c.
+// e.g. `Exec(true) == Ignore == Exec(false)` but `Exec(true) != Exec(false)`
+
+impl ExecFlag {
+    fn matches(&self, other: &Self) -> bool {
+        match (self, other) {
+            (ExecFlag::Exec(a), ExecFlag::Exec(b)) => a == b,
+            // Always treat as equal if either is Ignore.
+            _ => true,
+        }
+    }
+
+    /// Create a bool in an environment where we can't check a IgnoreExec value.
+    fn from_bool_unchecked(executable: bool) -> Self {
+        #[cfg(unix)]
+        let exec_flag = ExecFlag::Exec(executable);
+        #[cfg(windows)]
+        let exec_flag = ExecFlag::Ignore;
+        exec_flag
+    }
+}
+
+/// Whether to ignore the executable bit when comparing files. The executable
+/// state is always ignored on Windows, but is respected by default on Unix and
+/// is ignored if we find that the filesystem doesn't support it or by user
+/// configuration.
 #[cfg(unix)]
-type FileExecutableFlag = bool;
+#[derive(Debug, Clone, Copy)]
+struct IgnoreExec(bool);
 #[cfg(windows)]
-type FileExecutableFlag = ();
+#[derive(Debug, Clone, Copy)]
+struct IgnoreExec();
```

Comment:
nit: you can drop the parentheses, but it's also fine if you prefer to keep them